src/main/java/fi/aalto/cs/apluscourses/ui/base/BaseListView.java

```diff
@@ -91,6 +96,49 @@ public void setModel(@Nullable ListModel<E> model) {
     }
   }
 
+  /**
+  * Set the model of the this tree to the given view model, or do nothing if the given view model
+  * is {@code null}.
+  */
+  public void setViewModel(@Nullable BaseListViewModel<E> viewModel) {
+    if (viewModel != null) {
+      synchronized (viewModelLock) {
+        unregisterViewModel();
+        this.viewModel = viewModel;
+        registerViewModel();
+      }
+      update();
+      viewModel.getSelectionModel().clearSelection();
+    }
+  }
+
+  private void unregisterViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.removeCallback(this);
+      }
+    }
+  }
+
+  private void registerViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.addListener(this, BaseListView::update, SwingUtilities::invokeLater);
+      }
+    }
+  }
+
+  private void update() {
+    BaseListViewModel<E> localViewModel;
+    synchronized (viewModelLock) {
+      localViewModel = this.viewModel;
+    }
+    var listModel = new DefaultListModel<E>();
+    //keeping selection
+    listModel.addAll(localViewModel.streamVisibleItems().collect(Collectors.toList()));
+    setModel(listModel);
```

Comment:
I'm wondering if it would be good to include
```java
synchronized (viewModelLock) {
  if (localViewModel != this.viewModel) return;
}
```
before the last line. What do you think? And the same thing for the treeview. And could the code be reused between these two classes?